great_expectations/datasource/data_connector/util.py

```diff
@@ -333,6 +341,48 @@ def _walk_blob_hierarchy(name_starts_with: str) -> None:
     return path_list
 
 
+def list_gcs_keys(
+    gcs: storage.Client,
+    query_options: dict,
+    recursive: bool = False,
+) -> List[str]:
+    """
+    Utilizes the GCS connection object to retrieve blob names based on user-provided criteria.
+
+    For InferredAssetGCSDataConnector, we take `bucket_or_name` and `prefix` and search for files using RegEx at and below the level
+    specified by those parameters. However, for ConfiguredAssetGCSDataConnector, we take `bucket_or_name` and `prefix` and
+    search for files using RegEx only at the level specified by that bucket and prefix.
+
+    This restriction for the ConfiguredAssetGCSDataConnector is needed because paths on GCS are comprised not only the leaf file name
+    but the full path that includes both the prefix and the file name. Otherwise, in the situations where multiple data assets
+    share levels of a directory tree, matching files to data assets will not be possible due to the path ambiguity.
+
+    Args:
+        gcs (storage.Client): GCS connnection object responsible for accessing bucket
+        query_options (dict): GCS query attributes ("bucket_or_name", "prefix", "delimiter", "max_results")
+        recursive (bool): True for InferredAssetGCSDataConnector and False for ConfiguredAssetGCSDataConnector (see above)
+
+    Returns:
+        List of keys representing GCS file paths (as filtered by the `query_options` dict)
+    """
+    # Delimiter determines whether or not traversal of bucket is recursive
+    # Manually set to appropriate default if not already set by user
+    delimiter = query_options["delimiter"]
+    if delimiter is None and not recursive:
+        query_options["delimiter"] = "/"
+    elif delimiter is not None and recursive:
+        query_options["delimiter"] = None
```

Comment:
This is a little tricky because the value of `delimiter` directly impacts how traversal works in the bucket. If you set it to be `/`, you immediately stop traversal once you finish iterating through the current level.

What do we do if a user asks for a recursive usage but also explicitly passes in a delimiter of `/`? Do we overwrite what the user put in to ensure proper behavior? Or do we leave it as is and have the user get unexpected behavior?

Reply:
Do you think it would help to display a warning to the user if Is there a way that we can message a warning to the user? I'm thinking either a logger message or even a print statement

Reply:
This is a great idea! Do you think logging or using warning is appropriate here?

Reply:
That's a good question. between `logger.warning()` and `warnings.warn()` I'm leaning towards `warnings.warn()`, but only because we want to tell the user about potentially unexpected behavior.. but it's your call. I'd support either